File: src/dotnet/PloyBus.Tests/Transport/Transactions/Messages/MessagesTests.cs
using NUnit.Framework;

namespace PolyBus.Transport.Transactions.Messages;

[TestFixture]
public class MessagesTests
{
    private Messages _messages = null!;

    [SetUp]
    public void SetUp()
    {
        _messages = new Messages();
    }

    [Test]
    public void Add_ValidMessageType_ReturnsMessageInfo()
    {
        // Act
        var result = _messages.Add(typeof(Command));

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result.Type, Is.EqualTo(MessageType.Command));
        Assert.That(result.Endpoint, Is.EqualTo("polybus"));
        Assert.That(result.Name, Is.EqualTo("polybus-command"));
        Assert.That(result.Major, Is.EqualTo(1));
        Assert.That(result.Minor, Is.EqualTo(0));
        Assert.That(result.Patch, Is.EqualTo(0));
    }

    [Test]
    public void Add_MessageTypeWithoutAttribute_ThrowsError()
    {
        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.Add(typeof(MessageWithoutAttribute)));
    }

    [Test]
    public void GetMessageInfo_ExistingType_ReturnsCorrectMessageInfo()
    {
        // Arrange
        _messages.Add(typeof(Command));

        // Act
        var result = _messages.GetMessageInfo(typeof(Command));

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result.Type, Is.EqualTo(MessageType.Command));
        Assert.That(result.Endpoint, Is.EqualTo("polybus"));
        Assert.That(result.Name, Is.EqualTo("polybus-command"));
    }

    [Test]
    public void GetMessageInfo_NonExistentType_ThrowsError()
    {
        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.GetMessageInfo(typeof(Command)));
    }

    [Test]
    public void GetTypeByMessageInfo_ExistingMessageInfo_ReturnsCorrectType()
    {
        // Arrange
        _messages.Add(typeof(Event));
        var messageInfo = new MessageInfo(MessageType.Event, "polybus", "polybus-event", 2, 1, 3);

        // Act
        var result = _messages.GetTypeByMessageInfo(messageInfo);

        // Assert
        Assert.That(result, Is.EqualTo(typeof(Event)));
    }

    [Test]
    public void GetTypeByMessageInfo_NonExistentMessageInfo_ThrowsError()
    {
        // Arrange
        var messageInfo = new MessageInfo(MessageType.Command, "unknown", "unknown-command", 1, 0, 0);

        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.GetTypeByMessageInfo(messageInfo));
    }

    [Test]
    public void GetTypeByMessageInfo_DifferentMinorPatchVersions_ReturnsType()
    {
        // Arrange
        _messages.Add(typeof(Event)); // Has version 2.1.3
        var messageInfoDifferentMinor = new MessageInfo(MessageType.Event, "polybus", "polybus-event", 2, 5, 3);
        var messageInfoDifferentPatch = new MessageInfo(MessageType.Event, "polybus", "polybus-event", 2, 1, 9);

        // Act
        var result1 = _messages.GetTypeByMessageInfo(messageInfoDifferentMinor);
        var result2 = _messages.GetTypeByMessageInfo(messageInfoDifferentPatch);

        // Assert
        Assert.That(result1, Is.EqualTo(typeof(Event)));
        Assert.That(result2, Is.EqualTo(typeof(Event)));
    }

    [Test]
    public void GetTypeByMessageInfo_DifferentMajorVersion_ThrowsError()
    {
        // Arrange
        _messages.Add(typeof(Event)); // Has version 2.1.3
        var messageInfoDifferentMajor = new MessageInfo(MessageType.Event, "polybus", "polybus-event", 3, 1, 3);

        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.GetTypeByMessageInfo(messageInfoDifferentMajor));
    }

    [Test]
    public void Add_SameTypeTwice_ThrowsError()
    {
        // Arrange
        _messages.Add(typeof(Command));

        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.Add(typeof(Command)));
    }

    [Test]
    public void GetHeaderByMessageInfo_ExistingMessageInfo_ReturnsCorrectHeader()
    {
        // Arrange
        _messages.Add(typeof(Command));
        var messageInfo = new MessageInfo(MessageType.Command, "polybus", "polybus-command", 1, 0, 0);

        // Act
        var result = _messages.GetHeaderByMessageInfo(messageInfo);

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result, Is.Not.Empty);
        Assert.That(result, Is.EqualTo(messageInfo.ToString(true)));
    }

    [Test]
    public void GetHeaderByMessageInfo_NonExistentMessageInfo_ThrowsError()
    {
        // Arrange
        var messageInfo = new MessageInfo(MessageType.Command, "unknown", "unknown-command", 1, 0, 0);

        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.GetHeaderByMessageInfo(messageInfo));
    }

    [Test]
    public void GetHeaderByMessageInfo_DifferentMajorVersion_ThrowsError()
    {
        // Arrange
        _messages.Add(typeof(Event)); // Has version 2.1.3
        var messageInfoDifferentMajor = new MessageInfo(MessageType.Event, "polybus", "polybus-event", 3, 1, 3);

        // Act & Assert
        Assert.Throws<PolyBusMessageNotFoundError>(() => _messages.GetHeaderByMessageInfo(messageInfoDifferentMajor));
    }
}

File: src/kotlin/src/test/kotlin/polybus/transport/transactions/messages/MessagesTests.kt
package polybus.transport.transactions.messages

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class MessagesTests {
    private lateinit var messages: Messages

    @BeforeEach
    fun setUp() {
        messages = Messages()
    }

    @Test
    fun addValidMessageTypeReturnsMessageInfo() {
        // Act
        val result = messages.add(Command::class.java)

        // Assert
        assertNotNull(result)
        assertEquals(MessageType.COMMAND, result.type)
        assertEquals("polybus", result.endpoint)
        assertEquals("polybus-command", result.name)
        assertEquals(1, result.major)
        assertEquals(0, result.minor)
        assertEquals(0, result.patch)
    }

    @Test
    fun addMessageTypeWithoutAttributeThrowsError() {
        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.add(MessageWithoutAttribute::class.java) }
    }

    @Test
    fun getMessageInfoExistingTypeReturnsCorrectMessageInfo() {
        // Arrange
        messages.add(Command::class.java)

        // Act
        val result = messages.getMessageInfo(Command::class.java)

        // Assert
        assertNotNull(result)
        assertEquals(MessageType.COMMAND, result.type)
        assertEquals("polybus", result.endpoint)
        assertEquals("polybus-command", result.name)
    }

    @Test
    fun getMessageInfoNonExistentTypeThrowsError() {
        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.getMessageInfo(Command::class.java) }
    }

    @Test
    fun getTypeByMessageInfoExistingMessageInfoReturnsCorrectType() {
        // Arrange
        messages.add(Event::class.java)
        val messageInfo = MessageInfoValue(MessageType.EVENT, "polybus", "polybus-event", 2, 1, 3)

        // Act
        val result = messages.getTypeByMessageInfo(messageInfo)

        // Assert
        assertEquals(Event::class.java, result)
    }

    @Test
    fun getTypeByMessageInfoNonExistentMessageInfoThrowsError() {
        // Arrange
        val messageInfo = MessageInfoValue(MessageType.COMMAND, "unknown", "unknown-command", 1, 0, 0)

        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.getTypeByMessageInfo(messageInfo) }
    }

    @Test
    fun getTypeByMessageInfoDifferentMinorPatchVersionsReturnsType() {
        // Arrange
        messages.add(Event::class.java) // Has version 2.1.3
        val messageInfoDifferentMinor = MessageInfoValue(MessageType.EVENT, "polybus", "polybus-event", 2, 5, 3)
        val messageInfoDifferentPatch = MessageInfoValue(MessageType.EVENT, "polybus", "polybus-event", 2, 1, 9)

        // Act
        val result1 = messages.getTypeByMessageInfo(messageInfoDifferentMinor)
        val result2 = messages.getTypeByMessageInfo(messageInfoDifferentPatch)

        // Assert
        assertEquals(Event::class.java, result1)
        assertEquals(Event::class.java, result2)
    }

    @Test
    fun getTypeByMessageInfoDifferentMajorVersionThrowsError() {
        // Arrange
        messages.add(Event::class.java) // Has version 2.1.3
        val messageInfoDifferentMajor = MessageInfoValue(MessageType.EVENT, "polybus", "polybus-event", 3, 1, 3)

        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.getTypeByMessageInfo(messageInfoDifferentMajor) }
    }

    @Test
    fun addSameTypeTwiceThrowsError() {
        // Arrange
        messages.add(Command::class.java)

        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.add(Command::class.java) }
    }

    @Test
    fun getHeaderByMessageInfoExistingMessageInfoReturnsCorrectHeader() {
        // Arrange
        messages.add(Command::class.java)
        val messageInfo = MessageInfoValue(MessageType.COMMAND, "polybus", "polybus-command", 1, 0, 0)

        // Act
        val result = messages.getHeaderByMessageInfo(messageInfo)

        // Assert
        assertNotNull(result)
        assertFalse(result.isEmpty())
        assertEquals(messageInfo.toHeaderString(true), result)
    }

    @Test
    fun getHeaderByMessageInfoNonExistentMessageInfoThrowsError() {
        // Arrange
        val messageInfo = MessageInfoValue(MessageType.COMMAND, "unknown", "unknown-command", 1, 0, 0)

        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.getHeaderByMessageInfo(messageInfo) }
    }

    @Test
    fun getHeaderByMessageInfoDifferentMajorVersionThrowsError() {
        // Arrange
        messages.add(Event::class.java) // Has version 2.1.3
        val messageInfoDifferentMajor = MessageInfoValue(MessageType.EVENT, "polybus", "polybus-event", 3, 1, 3)

        // Act & Assert
        assertThrows<PolyBusMessageNotFoundError> { messages.getHeaderByMessageInfo(messageInfoDifferentMajor) }
    }
}
